crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+        match (self, search_path.into()) {
+            (Self::Extra(ExtraPath(path)), Self::Extra(_)) => db.file_system().is_directory(path),
+            (Self::FirstParty(FirstPartyPath(path)), Self::FirstParty(_)) => db.file_system().is_directory(path),
+            (Self::SitePackages(SitePackagesPath(path)), Self::SitePackages(_)) => db.file_system().is_directory(path),
+            (Self::StandardLibrary(StandardLibraryPath(path)), Self::StandardLibrary(stdlib_root)) => {
+                let Some(module_name) = self.as_module_name() else {
+                    return false;
+                };
+                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
+                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        db.file_system().is_directory(path)
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn is_regular_package(&self, db: &dyn Db, search_path: impl Into<Self>) -> bool {
+        match (self, search_path.into()) {
+            (Self::Extra(ExtraPath(fs_path)), Self::Extra(_))
+            | (Self::FirstParty(FirstPartyPath(fs_path)), Self::FirstParty(_))
+            | (Self::SitePackages(SitePackagesPath(fs_path)), Self::SitePackages(_)) => {
+                let file_system = db.file_system();
+                file_system.exists(&fs_path.join("__init__.py"))
+                    || file_system.exists(&fs_path.join("__init__.pyi"))
+            }
+            // Unlike the other variants:
+            // (1) Account for VERSIONS
+            // (2) Only test for `__init__.pyi`, not `__init__.py`
+            (Self::StandardLibrary(StandardLibraryPath(fs_path)), Self::StandardLibrary(stdlib_root)) => {
+                let Some(module_name) = self.as_module_name() else {
+                    return false;
+                };
+                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
+                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        db.file_system().exists(&fs_path.join("__init__.pyi"))
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
```

Comment:
Not a huge deal, but the method is non trivial and the `impl Into<Self>` means that this method will get monomorphized for each type; You can avoid monomorphization by creating an inline function
```suggestion
    pub(crate) fn is_regular_package(&self, db: &dyn Db, search_path: impl Into<Self>) -> bool {
    		fn is_regular_impl(path: &ModuleResolutionSearchPath, db: &dyn Db, search_path) -> bool {
	        match (path, search_path) {
	            (Self::Extra(ExtraPath(fs_path)), Self::Extra(_))
	            | (Self::FirstParty(FirstPartyPath(fs_path)), Self::FirstParty(_))
	            | (Self::SitePackages(SitePackagesPath(fs_path)), Self::SitePackages(_)) => {
	                let file_system = db.file_system();
	                file_system.exists(&fs_path.join("__init__.py"))
	                    || file_system.exists(&fs_path.join("__init__.pyi"))
	            }
	            // Unlike the other variants:
	            // (1) Account for VERSIONS
	            // (2) Only test for `__init__.pyi`, not `__init__.py`
	            (Self::StandardLibrary(StandardLibraryPath(fs_path)), Self::StandardLibrary(stdlib_root)) => {
	                let Some(module_name) = self.as_module_name() else {
	                    return false;
	                };
	                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
	                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
	                    TypeshedVersionsQueryResult::Exists
	                    | TypeshedVersionsQueryResult::MaybeExists => {
	                        db.file_system().exists(&fs_path.join("__init__.pyi"))
	                    }
	                    TypeshedVersionsQueryResult::DoesNotExist => false,
	                }
	            }
	            (path, root) => unreachable!(
	                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
	            )
	        }
	     }
	     is_regular_impl(self, db, search_path.into())    
    }
```